File: server_play_akka_scala/app/domain/logic/LunchResolverFeldkueche.scala
package domain.logic

import java.net.URL
import java.time.{DayOfWeek, LocalDate}
import java.time.format.DateTimeFormatter

import domain.models.{LunchOffer, LunchProvider}
import infrastructure.OcrClient
import org.htmlcleaner.{CleanerProperties, HtmlCleaner}
import org.joda.money.{CurrencyUnit, Money}

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.matching.Regex

/**
 * Mittagsangebote von Feldküche Rühlow ermitteln.
 */
class LunchResolverFeldkueche(
    dateValidator: DateValidator,
    ocrClient: OcrClient) extends LunchResolver {

  sealed abstract class Weekday(val name: String, val order: Int)

  object Weekday {
    case object MONTAG extends Weekday("Montag", 0)
    case object DIENSTAG extends Weekday("Dienstag", 1)
    case object MITTWOCH extends Weekday("Mittwoch", 2)
    case object DONNERSTAG extends Weekday("Donnerstag", 3)
    case object FREITAG extends Weekday("Freitag", 4)

    val weekdaysValues = List[Weekday](MONTAG, DIENSTAG, MITTWOCH, DONNERSTAG, FREITAG)
    // TODO: improve with macro, see https://github.com/d6y/enumeration-examples & http://underscore.io/blog/posts/2014/09/03/enumerations.html
    val values = weekdaysValues
  }

  override def resolve: Future[Seq[LunchOffer]] =
    Future { resolveImageLinks(new URL("http://www.feldkuechebkarow.de/speiseplan")) }
      .flatMap(imageUrls => resolveFromImageLinks(imageUrls))

  private[logic] def resolveImageLinks(htmlUrl: URL): Seq[URL] = {
    val props = new CleanerProperties
    props.setCharset("utf-8")

    val rootNode = new HtmlCleaner(props).clean(htmlUrl)
    val links = rootNode.evaluateXPath("//div[@id='content_area']//a/@href").map { case n: String => n }.toSet
    links.map { link => new URL(link) }.toSeq
  }

  private def resolveFromImageLinks(imageUrls: Seq[URL]): Future[Seq[LunchOffer]] = Future.sequence {
    imageUrls.map(imageUrl => resolveTextFromImageLink(imageUrl))
  }.map(_.flatten)

  private def resolveTextFromImageLink(imageUrl: URL): Future[Seq[LunchOffer]] =
    ocrClient.doOCR(imageUrl).map(resolveOffersFromText)

  private[logic] def resolveOffersFromText(rawText: String): Seq[LunchOffer] = {
    val contentAsLines = rawText.split('\n').map(correctOcrErrors)
    val mondayOpt =
      for (
        wochenplanZeile <- contentAsLines.find(_.contains("Wochenplan"));
        day <- parseDay(wochenplanZeile);
        monday = day.`with`(DayOfWeek.MONDAY) if dateValidator.isValid(monday)
      ) yield monday

    val weekdaysRegex = s"(${Weekday.values.map(_.name).mkString("|")}) .*"
    val offerTexts = contentAsLines.filter(_.matches(weekdaysRegex)).filterNot(_.startsWith("Montag bis"))
    val prices = contentAsLines.filter(_.matches(""".*\d{1,},(\d{2}) €""")).flatMap(parsePrice)

    for (
      (offer, price) <- offerTexts.zip(prices);
      (weekday, name) <- parseOffer(offer);
      monday <- mondayOpt
    ) yield LunchOffer(0, name, monday.plusDays(weekday.order), price, LunchProvider.FELDKUECHE.id)
  }

  private def parseOffer(text: String): Option[(Weekday, String)] = {
    val (weekdayName, lunchName) = text.span(_ != ' ')
    Weekday.values.find(_.name == weekdayName) match {
      case Some(weekday) => Some(weekday, lunchName.replaceAll(""" \d{1,},\d{2} €""", "").trim)
      case _ => None
    }
  }

  private def correctOcrErrors(line: String) =
    line.trim
      .replaceAll("‚", ",")
      .replaceAll("""l\)""", "0")
      .replaceAll("""(\d{1,},\d{2}) [cCeE]""", "$1 €")
      .replaceAll("4,00 €", "4,80 €") // besser falsch zu hohe 4,80 als zu niedrige 4,00
      .replaceAll("Kanonen-", "Kartoffeln")
      .replaceAll("ﬂ", "fl")
      .replaceAll("ﬁ", "fi")
      .replaceAll("Wochen.lan", "Wochenplan")
      .replaceAll("Wochen.Ian", "Wochenplan")
      .replaceAll("eintopl", "eintopf")
      .replaceAll("eintop", "eintopf")
      .replaceAll("eintof", "eintopf")
      .replaceAll("eintopff", "eintopf")
      .replaceAll("Erhsen", "Erbsen")
      .replaceAll("Chili con Garne", "Chili con Carne")
      .replaceAll("—", "-")

  implicit class RegexContext(sc: StringContext) {
    def r = new Regex(sc.parts.mkString, sc.parts.tail.map(_ => "x"): _*)
  }

  /**
   * Erzeugt ein LocalDate aus dem Format "*dd.mm.yyyy*"
   *
   * @param text auszuwertender Text
   * @return
   */
  private def parseDay(text: String): Option[LocalDate] = text match {
    case r""".*(\d{2}\.\d{2}\.\d{2})$dayString.*""" => parseLocalDate(dayString, "dd.MM.yy")
    case _ => None
  }

  /**
   * Erzeugt ein Money-Objekt (in EURO) aus dem Format "*0,00*"
   *
   * @param text auszuwertender Text
   * @return
   */
  private def parsePrice(text: String): Option[Money] = text match {
    case r""".*(\d{1,})$major,(\d{2})$minor €""" => Some(Money.ofMinor(CurrencyUnit.EUR, major.toInt * 100 + minor.toInt))
    case _ => None
  }

  private def parseLocalDate(dateString: String, dateFormat: String): Option[LocalDate] =
    try {
      Some(LocalDate.from(DateTimeFormatter.ofPattern(dateFormat).parse(dateString)))
    } catch {
      case exc: Throwable => None
    }

}

File: backend-spring-kotlin/src/main/kotlin/lunchbox/domain/resolvers/LunchResolverFeldkueche.kt
package lunchbox.domain.resolvers

import lunchbox.domain.models.LunchOffer
import lunchbox.domain.models.LunchProvider.FELDKUECHE
import lunchbox.util.date.DateValidator
import lunchbox.util.html.HtmlParser
import lunchbox.util.ocr.OcrClient
import lunchbox.util.string.StringParser
import lunchbox.util.url.UrlUtil.url
import org.joda.money.Money
import org.springframework.stereotype.Component
import java.net.URL
import java.time.DayOfWeek
import java.time.LocalDate

/**
 * Mittagsangebote von Feldküche Rühlow ermitteln.
 */
@Component
class LunchResolverFeldkueche(
  val dateValidator: DateValidator,
  val ocrClient: OcrClient,
  val htmlParser: HtmlParser,
) : LunchResolver {
  override val provider = FELDKUECHE

  override fun resolve(): List<LunchOffer> =
    // TODO: parallelize
    resolveFromImageLinks(resolveImageLinks(provider.menuUrl))

  fun resolveImageLinks(htmlUrl: URL): List<URL> {
    val site = htmlParser.parse(htmlUrl)

    val divsWithContent = site.select("div#content_area")
    val links = divsWithContent.select("a").map { it.attr("href") }
    return links.map { url(it) }
  }

  private fun resolveFromImageLinks(imageUrls: List<URL>): List<LunchOffer> =
    imageUrls.flatMap { resolveTextFromImageLink(it) }

  private fun resolveTextFromImageLink(imageUrl: URL): List<LunchOffer> =
    resolveOffersFromText(ocrClient.doOCR(imageUrl))

  fun resolveOffersFromText(rawText: String): List<LunchOffer> {
    val contentAsLines = rawText.split('\n').map { correctOcrErrors(it) }
    val monday = resolveMonday(contentAsLines) ?: return emptyList()

    val rawOffers =
      if (Weekday.entries.any { contentAsLines.contains(it.label) } &&
        contentAsLines.any { it.matches(Regex("""^\d+,(\d{2}) *€$""")) }
      ) {
        resolveOffersWith3rowSplit(contentAsLines)
      } else {
        resolveOffersWith2rowSplit(contentAsLines)
      }

    return rawOffers
      .map { createOffer(it, monday) }
      .filter { dateValidator.isValid(it.day, provider) }
  }

  private fun createOffer(
    raw: RawOffer,
    monday: LocalDate,
  ): LunchOffer {
    val (title, description) =
      StringParser.splitOfferName(
        raw.name,
        listOf(" mit ", "Kartoffeln", "Brot", "Nudeln", "Klöße"),
      )

    return LunchOffer(
      0,
      title,
      description,
      monday.plusDays(raw.weekday.order),
      raw.price,
      emptySet(),
      provider.id,
    )
  }

  private fun resolveMonday(contentAsLines: List<String>): LocalDate? {
    val wochenplanZeile = contentAsLines.find { it.contains("Wochenplan") } ?: return null
    val day = StringParser.parseLocalDate(wochenplanZeile) ?: return null
    val monday = day.with(DayOfWeek.MONDAY)
    return if (dateValidator.isValid(monday, provider)) monday else null
  }

  private fun resolveOffersWith2rowSplit(contentAsLines: List<String>): List<RawOffer> {
    val weekdaysRegex = Regex("(${Weekday.entries.joinToString("|"){ it.label }}) .*")
    val offerTexts =
      contentAsLines
        .filter { it.matches(weekdaysRegex) }
        .filterNot { it.startsWith("Montag bis") }

    fun splitWeekdayAndName(text: String): RawOfferName? {
      val weekdayName = text.takeWhile { it != ' ' }
      val lunchName = text.dropWhile { it != ' ' }
      val weekday = Weekday.entries.find { it.label == weekdayName } ?: return null

      return RawOfferName(weekday, lunchName.replace(Regex(""" \d+,\d{2} *€"""), "").trim())
    }

    val prices =
      contentAsLines
        .filter { it.matches(Regex(""".*\d+,(\d{2}) *€""")) }
        .mapNotNull { StringParser.parseMoney(it) }

    return offerTexts
      .mapNotNull { splitWeekdayAndName(it) }
      .zip(prices)
      .map { (rawName, price) -> RawOffer(rawName.weekday, rawName.name, price) }
  }

  private fun resolveOffersWith3rowSplit(contentAsRows: List<String>): List<RawOffer> {
    fun isWeekday(text: String) = Weekday.entries.map { it.label }.contains(text)
    val relevantRows =
      contentAsRows
        .filter { it.isNotEmpty() }
        .dropWhile { !isWeekday(it) }
    val weekdayTextRows = relevantRows.takeWhile { isWeekday(it) }
    val followingRows = relevantRows.dropWhile { isWeekday(it) }

    val weekdayRows =
      weekdayTextRows
        .mapNotNull { row -> Weekday.entries.find { it.label == row } }

    val prices =
      followingRows
        .filter { it.matches(Regex(""".*\d+,(\d{2}) *€""")) }
        .mapNotNull { StringParser.parseMoney(it) }

    return weekdayRows
      .zip(followingRows)
      .map { (weekday, name) -> RawOfferName(weekday, name) }
      .zip(prices)
      .map { (rawName, price) -> RawOffer(rawName.weekday, rawName.name, price) }
  }

  private fun correctOcrErrors(line: String) =
    line
      .trim()
      .replace("Wochenglan", "Wochenplan")
      .replace("‚", ",")
      .replace("""l)""", "0")
      .replace(Regex("""(\d+,\d{2}) [cCeEs]"""), "$1 €")
      //      .replace("4,00 €", "4,80 €") // besser falsch zu hohe 4,80 als zu niedrige 4,00
      .replace("Kanonen-", "Kartoffeln")
      .replace("ﬂ", "fl")
      .replace("ﬁ", "fi")
      .replace(Regex("Wochen.lan"), "Wochenplan")
      .replace(Regex("Wochen.Ian"), "Wochenplan")
      .replace("eintopl", "eintopf")
      .replace("eintop", "eintopf")
      .replace("eintof", "eintopf")
      .replace("eintopff", "eintopf")
      .replace("Erhsen", "Erbsen")
      .replace("Chili con Garne", "Chili con Carne")
      .replace("—", "-")
      .replace(" _", " ")

  data class RawOfferName(
    val weekday: Weekday,
    val name: String,
  )

  data class RawOffer(
    val weekday: Weekday,
    val name: String,
    val price: Money,
  )

  enum class Weekday(
    val label: String,
    val order: Long,
  ) {
    MONTAG("Montag", 0),
    DIENSTAG("Dienstag", 1),
    MITTWOCH("Mittwoch", 2),
    DONNERSTAG("Donnerstag", 3),
    FREITAG("Freitag", 4),
  }
}
